### ISSUE TITLE
Remove (or noop) all client facing freeze operations

### ISSUE THREAD

**Issue by gregscullard** [hashgraph/hedera-sdk-js#1445]
### Problem

Freezing a transaction using the SDK is currently necessary in order to serialise it.
It should be possible to serialise a transaction without freezing it and in addition, it should be possible to serialise an incomplete transaction (e.g. missing node_account_id and transaction_id as a minimum, maybe others).
Below is the rationale.

* The SDK manages a list of nodes and marks unreachable nodes when it encounters one.
* Dapps using wallets such as Hashpack construct a transaction using the SDK and need to serialise it in order to submit it to the wallet for signing
* This requires the dapp to set the TransactionId and NodeAccountId in order to freeze the transaction to serialise
* The wallet signs and submits, given the node accountId is set, the SDK cannot override it.
* The dApp SDK isn't submitting to the network, so it is not aware of nodes that are unavailable and may keep selecting the same incorrect node

The freeze feature in itself doesn't appear to add much value in any case, failure to complete a transaction before serialisation or execution will result an error that would be caught in testing. In fact, as a developer it's more confusing and unnecessary additional coding steps and complexity.

### Solution

Remove the need to freeze a transaction, suggest deprecating and making the method a no operation during the deprecation period.

Allow any transaction to be serialised/de-serialised, the receiver or the serialised transaction can fill in the blanks before submitting (likely done automatically by the sdk at .execute()).

HIP: https://hips.hedera.com/hip/hip-745
### Alternatives

The alternative is to keep the SDK as it is.
```[tasklist]
### Tasks
- [x] Write a HIP
- [ ] Create a issue in the references repo and propose a design for the SDKs in the references repo
```

**Comment by petreze** [comment]
Hello @gregscullard 
I had a look on the whole concept of freezing transactions. I can see that upon calling `freezeWith(client)`, the SDK sets most of the fields for the following transaction. All of them are dependent on the passed `client` (some of them are derived directly from it, others like `nodeAccountIds` need the client in order to calculate/extract the needed value to be set)

I can agree that the whole freezing necessity is somewhat confusing as a developer and adds unnecessary complexity. But if we want to remove it, one approach could be to introduce a `client` variable to every exit condition method **i.e. toBytes[Async]()`, `getTransactionHash[PerNode]()** in order to perform them without the need for freezing the transaction before that.

Also some of the locking mechanisms should be removed in order for some of the fields to be able to be set later on, after some serialization/deserialization processes

**Comment by gregscullard** [comment]
The challenge here is that I could technically have a software solution that doesn't have a "client", but is preparing transactions, serialises them and passes them onto another application that has a "client" and is able to fill in the missing information, sign, etc...

The SDK as it is designed assumes a client exists, which is not necessarily the case.

Imo, freeze is trying to ensure a serialised transaction is complete and ready for signing which is great, but it prevents some use cases from being implemented or enforces strange development behaviour (e.g. creating a client with a fake account/private key in order to be able to freeze in the first place) and while it's trying to help developers not make mistakes (or make changes to transactions that already have signatures), it's too intrusive imo.

**Comment by petreze** [comment]
@gregscullard 
I see your concerns. However, currently the SDK needs the `nodeAccountIds` in order to sign the transactions. As we were discussing the possibility to remove the `nodeAccountId` from the signing process but for now that is not the case and we should include it.
There are some things that come up after this fact:

- In my opinion we cannot provide a functionality to serialise an incomplete transaction (e.g. missing node_account_id) because we need the `nodeAccountIds` list to be filled in order to sign the transactions.
- Basically this means that either the node against which we will try to execute the transaction has to be specified via `Transaction.setNodeAccountIds()` or  the **client** has to be provided to extract and set `nodeAccountIds` by calling `(client._network.getNodeAccountIdsForExecute())`

**Comment by gregscullard** [comment]
I agree `node_account_ids` are needed to sign the transaction, however if I'm a dapp generating a transaction which is signed by a wallet, the dApp isn't signing the transaction, it's sending it to the wallet to sign, but currently needs to provide `node_account_ids` in order to serialise an unsigned transaction...

Whatever freeze does could be moved to .execute() or .sign(), allowing an incomplete transaction to be serialised or de-serialised so it can be transferred onto the wallet (or whoever needs it), they will be responsible for completing missing data (automatically via .execute() or manually if desired).

This is how it used to work, .freeze() was introduced for serialisation purposes, but I still don't understand what value it's really providing other than preventing an incomplete transaction from being serialised and adding an unnecessary burden on developers. Serialising imo should just be there to enable transport of a transaction over the wires (in a JSON payload or other), it shouldn't be trying to validate the object being serialised.

**Comment by gregscullard** [comment]
If, for some reason the transaction is incomplete when sent to Hedera, this will be found very quickly in testing, the additional burden on the developer isn't warranted imo.

**Comment by petreze** [comment]
I agree. Will do a little bit of research of how we can approach this issue and refactor the hole process around freezing and signing

**Comment by petreze** [comment]
Hello @gregscullard 
I did a bit more work here and encountered one issue that might appear to be a blocker. This only happens whenever we have a transaction which has to be signed explicitly by other account than the one executing and paying for it. Since the goal is to make it possible for a transaction to be serialized/deserialized without the need for the `nodeAccountIds` and `transactionId` to be set and since the `transaction body` has to include the `nodeAccountId`, we are forced to pass a `null` value as a `nodeAccountId`, hence we are not going to have the actual transaction bytes that we can execute and send to the network.

Let's say the next thing we do is to generate the transaction again from those `transaction bytes` with `fromBytes()`. Then the `nodeAccountId` will be missing which is normal but we won't have access to the private key signing function. This is crucial because we would have to sign again the transaction, this time with specified `nodeAccountId` or with randomly chosen from the addressbook of the client's network.

Entire flow of `toBytes()` and `fromBytes()` happens with the help of the [protobufs](https://github.com/hashgraph/hedera-protobufs/blob/main/services/transaction.proto)

**Comment by SimiHunjan** [comment]
Noting this feature requires a HIP to be published to get feedback from applications before releasing any feature.

**Comment by gregscullard** [comment]
If I am a server and I generate a transaction that I sign and pass onto a wallet to counter-sign, I'll set the node account id and share the transaction bytes with the wallet to sign, it will then execute the tx, or return the signature to me.
If I am a dapp, I don't need to set the node account id, the wallet signing the transaction can take care of it.

**Comment by gregscullard** [comment]
If node account id is missing on signing a transaction, throw an exception, but don't throw if I try to serialise the transaction with node account id missing.

**Comment by SimiHunjan** [comment]
This has been completed via https://github.com/hashgraph/hedera-sdk-js/issues/1807
